core/src/smartcontracts/isi/asset.rs

```diff
@@ -54,89 +57,93 @@ pub mod isi {
+            Err: From<Error>,
+        {
+            let asset_id = mint.destination_id;
+
+            assert_can_mint(
+                &asset_id.definition_id,
+                wsv,
+                <Self as AssetInstructionInfo>::EXPECTED_VALUE_TYPE,
+            )?;
+            wsv.asset_or_insert(
+                &asset_id,
+                <Self as AssetInstructionInfo>::DEFAULT_ASSET_VALUE,
+            )?;
             wsv.modify_asset(&asset_id, |asset| {
-                let quantity: &mut u128 = asset
+                let quantity: &mut Self = asset
                     .try_as_mut()
                     .map_err(eyre::Error::from)
                     .map_err(|e| Error::Conversion(e.to_string()))?;
                 *quantity = quantity
-                    .checked_add(self.object)
+                    .checked_add(mint.object)
                     .ok_or(MathError::Overflow)?;
-                #[allow(clippy::cast_precision_loss)]
-                wsv.metrics.tx_amounts.observe(*quantity as f64);
+                wsv.metrics
+                    .tx_amounts
+                    .observe(<Self as FloatMetricFrom>::from(*quantity));
 
                 Ok(AssetEvent::Added(asset_id.clone()))
-            })
+            })?;
+            Ok(())
         }
     }
 
-    impl<W: WorldTrait> Execute<W> for Mint<Asset, Fixed> {
-        type Error = Error;
+    macro_rules! impl_mint {
+        ($ty:ty, $metrics:literal) => {
+            impl InnerMint for $ty {}
 
-        #[metrics(+"mint_fixed")]
-        fn execute(
-            self,
-            _authority: <Account as Identifiable>::Id,
-            wsv: &WorldStateView<W>,
-        ) -> Result<(), Self::Error> {
-            let asset_id = self.destination_id;
-
-            assert_can_mint(&asset_id.definition_id, wsv, AssetValueType::Fixed)?;
-            wsv.asset_or_insert(&asset_id, Fixed::ZERO)?;
-            wsv.modify_asset(&asset_id, |asset| {
-                let quantity: &mut Fixed = asset
-                    .try_as_mut()
-                    .map_err(eyre::Error::from)
-                    .map_err(|e| Error::Conversion(e.to_string()))?;
-                *quantity = quantity.checked_add(self.object)?;
-                wsv.metrics.tx_amounts.observe((*quantity).into());
+            impl<W: WorldTrait> Execute<W> for Mint<Asset, $ty> {
+                type Error = Error;
 
-                Ok(AssetEvent::Added(asset_id.clone()))
-            })
-        }
+                #[metrics(+$metrics)]
+                fn execute(
+                    self,
+                    authority: AccountId,
+                    wsv: &WorldStateView<W>,
+                ) -> Result<(), Self::Error> {
+                    <$ty as InnerMint>::execute(self, authority, wsv)
+                }
+            }
+        };
     }
 
+    impl_mint!(u32, "mint_qty");
+    impl_mint!(u128, "mint_big_qty");
+    impl_mint!(Fixed, "mint_fixed");
```

Comment:
Hm, If we had a way to use metrics without macro `metrics` we would be able to implement mint without this macro

Reply:
After that leave a link here, please